File: packages/sdk/src/runtime.ts
// The Axon agent runtime — the batteries-included worker.
//
// The low-level client gives you the primitives (register, getTaskHistory,
// startTask, completeTask, failTask, emitProgress). This wires them into a live
// agent: register once, then poll for queued work, run your handler, and settle
// each task — with concurrency, progress, graceful shutdown, and self-healing
// error handling. Building an earning agent goes from ~40 lines of glue to:
//
//   const agent = defineAgent(axon, {
//     agentId: "my-agent", name: "My Agent", capabilities: ["research"],
//     publicKey, walletAddress,
//     handler: async ({ task, progress }) => {
//       await progress("thinking…");
//       return await doTheWork(task.task);
//     },
//   });
//   await agent.start();

import { AxonApiError, AxonClient } from "./client";
import type { AgentContext, AgentRuntimeOptions, AxonAgent, TaskRequest } from "./types";

const sleep = (ms: number) => new Promise<void>((r) => setTimeout(r, ms));

function isNotFound(err: unknown): boolean {
  return err instanceof AxonApiError && (err.status === 404 || err.code === "NOT_FOUND");
}

// A task claimed by another worker between our poll and our startTask — expected
// under concurrency, not an error worth surfacing.
function isStateConflict(err: unknown): boolean {
  return err instanceof AxonApiError && (err.status === 409 || err.code === "TASK_STATE_CONFLICT");
}

/**
 * Define a long-running Axon agent. Returns a controller — call `start()` to
 * register (if needed) and begin processing queued tasks, `stop()` to drain and
 * shut down. The handler runs once per incoming task; return its output string
 * (or `{ output, success:false }` / throw to fail the task).
 */
export function defineAgent(client: AxonClient, options: AgentRuntimeOptions): AxonAgent {
  const {
    handler,
    pollIntervalMs = 2000,
    autoRegister = true,
    concurrency = 1,
    onError,
    onTaskStart,
    onTaskComplete,
    ...registration
  } = options;
  const agentId = registration.agentId;

  let running = false;
  let stopping = false;
  let loopPromise: Promise<void> | null = null;
  const inFlight = new Set<Promise<void>>();
  // Task ids currently being handled. A queued task can be returned by another
  // poll before our startTask transitions it out of `queued`; tracking claims
  // stops us launching the same task twice under concurrency > 1.
  const claiming = new Set<string>();

  async function ensureRegistered(): Promise<void> {
    if (!autoRegister) return;
    try {
      await client.getAgent(agentId);
    } catch (err) {
      if (isNotFound(err)) {
        await client.register(registration);
        return;
      }
      throw err;
    }
  }

  // Invoke a user lifecycle hook without letting a throw from it derail the
  // runtime (it would otherwise surface as an unhandled rejection on the detached
  // runOne promise).
  function safeCall<A extends unknown[]>(fn: ((...args: A) => void) | undefined, ...args: A): void {
    if (!fn) return;
    try {
      fn(...args);
    } catch {
      /* a throwing hook is the caller's bug, not ours to propagate */
    }
  }

  // Settle a finished task, retrying a few times so a transient backend blip
  // doesn't strand completed work in `running`. Returns whether it settled; on a
  // sustained failure the output is unrecoverable, so we surface it via onError.
  async function settle(started: TaskRequest, ok: boolean, text: string): Promise<boolean> {
    const attempts = 4;
    for (let i = 0; i < attempts; i++) {
      try {
        if (ok) await client.completeTask(started.taskId, text);
        else await client.failTask(started.taskId, text);
        return true;
      } catch (err) {
        // Task is already terminal — an earlier settle landed but its response was
        // lost, and the retry now sees a state conflict. That's success, not an
        // orphan: the task did settle.
        if (isStateConflict(err)) return true;
        if (i === attempts - 1) {
          safeCall(onError, err, started);
          return false;
        }
        await sleep(Math.min(2000, 200 * 2 ** i));
      }
    }
    return false;
  }

  async function runOne(task: TaskRequest): Promise<void> {
    // Claim the task. If another worker beat us to it, quietly move on.
    let started: TaskRequest;
    try {
      started = await client.startTask(task.taskId);
    } catch (err) {
      if (isStateConflict(err)) return;
      safeCall(onError, err, task);
      return;
    }

    safeCall(onTaskStart, started);
    const ctx: AgentContext = {
      task: started,
      // Progress is best-effort telemetry — a failed emit must never fail the
      // task the handler is otherwise completing fine.
      progress: (message: string) =>
        client.emitProgress(started.taskId, message).then(
          () => undefined,
          () => undefined,
        ),
      get stopping() {
        return stopping;
      },
    };

    // Run the handler → normalize to a terminal outcome. `text` is the output
    // when successful, the error message otherwise. A throw fails the task.
    let ok: boolean;
    let text: string;
    try {
      const result = await handler(ctx);
      if (typeof result === "string") {
        ok = true;
        text = result;
      } else {
        ok = result.success !== false;
        text = ok ? result.output : result.output || "Task failed";
      }
    } catch (err) {
      ok = false;
      text = err instanceof Error ? err.message : String(err);
      safeCall(onError, err, started);
    }

    const settled = await settle(started, ok, text);
    if (settled) {
      safeCall(onTaskComplete, {
        taskId: started.taskId,
        success: ok,
        output: ok ? text : "",
        error: ok ? undefined : text,
        completedAt: new Date().toISOString(),
      });
    }
  }

  async function loop(): Promise<void> {
    while (running) {
      let launched = 0;
      try {
        const slots = concurrency - inFlight.size;
        if (slots > 0) {
          const queued = await client.getTaskHistory({ agentId, role: "recipient", status: "queued", limit: slots });
          for (const task of queued) {
            if (!running) break;
            if (claiming.has(task.taskId)) continue; // already picked up this pass/loop
            claiming.add(task.taskId);
            const p = runOne(task).finally(() => {
              inFlight.delete(p);
              claiming.delete(task.taskId);
            });
            inFlight.add(p);
            launched++;
          }
        }
      } catch (err) {
        // A transient poll failure must not kill the loop.
        safeCall(onError, err);
      }
      // If we filled work this pass, loop again immediately to keep slots busy;
      // otherwise idle for the poll interval.
      if (launched === 0) await sleep(pollIntervalMs);
    }
  }

  return {
    get agentId() {
      return agentId;
    },
    get running() {
      return running;
    },
    async start() {
      if (running) return;
      // Set running synchronously — before the first await — so a second
      // concurrent start() is a genuine no-op and can't spawn a rival loop.
      running = true;
      stopping = false;
      try {
        await ensureRegistered();
      } catch (err) {
        running = false; // startup failed; allow another attempt
        throw err;
      }
      loopPromise = loop();
    },
    async stop() {
      stopping = true;
      running = false;
      // Wait for in-flight handlers to settle so no task is left half-done.
      await Promise.allSettled([...inFlight]);
      if (loopPromise) await loopPromise;
      loopPromise = null;
    },
  };
}

File: packages/sdk-python/axon/runtime.py
"""
The Axon agent runtime — the batteries-included worker.

Register a handler and the runtime polls for queued work, runs it, and settles
each task, on a background thread. Building an earning agent goes from wiring the
task primitives by hand to:

    agent = define_agent(
        client,
        agent_id="my-agent", name="My Agent", capabilities=["research"],
        public_key=pk, wallet_address=wallet,
        handler=lambda ctx: do_work(ctx.task["task"]),
    )
    agent.start()
    ...
    agent.stop()
"""

import threading
import time
from concurrent.futures import Future, ThreadPoolExecutor
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Any, Callable, Dict, List, Optional, Union

from .client import AxonClient
from .errors import AxonApiError

# handler(ctx) -> output string, or {"output": str, "success": bool}
HandlerResult = Union[str, Dict[str, Any]]
Handler = Callable[["AgentContext"], HandlerResult]


def _is_not_found(err: Exception) -> bool:
    return isinstance(err, AxonApiError) and (err.status == 404 or err.code == "NOT_FOUND")


def _is_state_conflict(err: Exception) -> bool:
    return isinstance(err, AxonApiError) and (err.status == 409 or err.code == "TASK_STATE_CONFLICT")


def _now() -> str:
    return datetime.now(timezone.utc).isoformat()


@dataclass
class AgentContext:
    task: Dict[str, Any]
    _agent: "AxonAgent"

    def progress(self, message: str) -> None:
        """Emit an intermediate progress message — best-effort, never fails the task."""
        try:
            self._agent.client.emit_progress(self.task["taskId"], message)
        except Exception:  # noqa: BLE001
            pass

    @property
    def stopping(self) -> bool:
        return self._agent._stopping


class AxonAgent:
    def __init__(
        self,
        client: AxonClient,
        registration: Dict[str, Any],
        handler: Handler,
        poll_interval_seconds: float = 2.0,
        auto_register: bool = True,
        concurrency: int = 1,
        on_error: Optional[Callable[..., None]] = None,
        on_task_start: Optional[Callable[[Dict[str, Any]], None]] = None,
        on_task_complete: Optional[Callable[[Dict[str, Any]], None]] = None,
    ):
        self.client = client
        self.registration = registration
        self.agent_id = registration["agentId"]
        self.handler = handler
        self.poll_interval = poll_interval_seconds
        self.auto_register = auto_register
        self.concurrency = max(1, concurrency)
        self.on_error = on_error
        self.on_task_start = on_task_start
        self.on_task_complete = on_task_complete

        self._running = False
        self._stopping = False
        self._thread: Optional[threading.Thread] = None
        self._executor = ThreadPoolExecutor(max_workers=self.concurrency)
        self._inflight: "set[Future]" = set()
        self._claiming: "set[str]" = set()
        self._lock = threading.Lock()

    @property
    def running(self) -> bool:
        return self._running

    def start(self) -> None:
        with self._lock:  # atomic check-and-set — a concurrent start() is a no-op
            if self._running:
                return
            self._running = True
            # Fresh executor each run so the agent is restartable after stop()
            # (stop() shuts the previous one down).
            self._executor = ThreadPoolExecutor(max_workers=self.concurrency)
        self._stopping = False
        try:
            self._ensure_registered()
        except Exception:
            self._running = False
            raise
        self._thread = threading.Thread(target=self._loop, daemon=True)
        self._thread.start()

    def stop(self) -> None:
        self._stopping = True
        self._running = False
        if self._thread:
            self._thread.join()
        with self._lock:
            pending = list(self._inflight)
        for fut in pending:  # drain in-flight handlers before shutting down
            try:
                fut.result()
            except Exception:  # noqa: BLE001
                pass
        self._executor.shutdown(wait=True)

    # ── internals ────────────────────────────────────────────────────────────
    def _safe(self, fn: Optional[Callable], *args: Any) -> None:
        if fn is None:
            return
        try:
            fn(*args)
        except Exception:  # noqa: BLE001
            pass

    def _ensure_registered(self) -> None:
        if not self.auto_register:
            return
        try:
            self.client.get_agent(self.agent_id)
        except Exception as err:
            if _is_not_found(err):
                self.client.register(self.registration)
                return
            raise

    def _loop(self) -> None:
        while self._running:
            launched = 0
            try:
                with self._lock:
                    slots = self.concurrency - len(self._inflight)
                if slots > 0:
                    queued: List[Dict[str, Any]] = self.client.get_task_history(
                        self.agent_id, role="recipient", status="queued", limit=slots
                    )
                    for t in queued:
                        if not self._running:
                            break
                        tid = t["taskId"]
                        with self._lock:
                            if tid in self._claiming or len(self._inflight) >= self.concurrency:
                                continue
                            self._claiming.add(tid)
                            fut = self._executor.submit(self._run_one, t)
                            self._inflight.add(fut)
                        # add_done_callback OUTSIDE the lock: on an already-finished
                        # future it runs _on_done synchronously, which needs the lock.
                        fut.add_done_callback(lambda f, tid=tid: self._on_done(f, tid))
                        launched += 1
            except Exception as err:  # noqa: BLE001
                self._safe(self.on_error, err, None)
            if launched == 0:
                time.sleep(self.poll_interval)

    def _on_done(self, fut: Future, task_id: str) -> None:
        with self._lock:
            self._inflight.discard(fut)
            self._claiming.discard(task_id)

    def _settle(self, task_id: str, ok: bool, text: str) -> bool:
        """Settle with a bounded retry; a state conflict means an earlier settle
        already landed (lost response), so treat it as success."""
        attempts = 4
        for i in range(attempts):
            try:
                if ok:
                    self.client.complete_task(task_id, text)
                else:
                    self.client.fail_task(task_id, text)
                return True
            except Exception as err:  # noqa: BLE001
                if _is_state_conflict(err):
                    return True
                if i == attempts - 1:
                    self._safe(self.on_error, err, {"taskId": task_id})
                    return False
                time.sleep(min(2.0, 0.2 * (2 ** i)))
        return False

    def _run_one(self, task: Dict[str, Any]) -> None:
        try:
            started = self.client.start_task(task["taskId"])
        except Exception as err:  # noqa: BLE001
            if _is_state_conflict(err):
                return
            self._safe(self.on_error, err, task)
            return

        self._safe(self.on_task_start, started)
        ctx = AgentContext(task=started, _agent=self)

        try:
            result = self.handler(ctx)
            if isinstance(result, str):
                ok, text = True, result
            else:
                ok = result.get("success", True) is not False
                text = result.get("output", "") if ok else (result.get("output") or "Task failed")
        except Exception as err:  # noqa: BLE001
            ok, text = False, str(err)
            self._safe(self.on_error, err, started)

        settled = self._settle(started["taskId"], ok, text)
        if settled:
            self._safe(self.on_task_complete, {
                "taskId": started["taskId"],
                "success": ok,
                "output": text if ok else "",
                "error": None if ok else text,
                "completedAt": _now(),
            })


def define_agent(
    client: AxonClient,
    handler: Handler,
    agent_id: str,
    name: str,
    capabilities: List[str],
    public_key: str,
    wallet_address: Optional[str] = None,
    price: Optional[str] = None,
    provider: Optional[str] = None,
    poll_interval_seconds: float = 2.0,
    auto_register: bool = True,
    concurrency: int = 1,
    on_error: Optional[Callable[..., None]] = None,
    on_task_start: Optional[Callable[[Dict[str, Any]], None]] = None,
    on_task_complete: Optional[Callable[[Dict[str, Any]], None]] = None,
    **extra_registration: Any,
) -> AxonAgent:
    """Define a long-running Axon agent. Call ``start()`` to register (if needed)
    and begin processing queued tasks, ``stop()`` to drain and shut down."""
    registration: Dict[str, Any] = {
        "agentId": agent_id,
        "name": name,
        "capabilities": capabilities,
        "publicKey": public_key,
    }
    if wallet_address:
        registration["walletAddress"] = wallet_address
    if price:
        registration["price"] = price
    if provider:
        registration["provider"] = provider
    registration.update(extra_registration)

    return AxonAgent(
        client, registration, handler,
        poll_interval_seconds=poll_interval_seconds,
        auto_register=auto_register, concurrency=concurrency,
        on_error=on_error, on_task_start=on_task_start, on_task_complete=on_task_complete,
    )
